pkg/portal/service/collaborator.go

```diff
@@ -138,6 +158,178 @@ func (s *CollaboratorService) DeleteCollaborator(c *model.Collaborator) error {
+	rows, err := s.SQLExecutor.QueryWith(q)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var is []*model.CollaboratorInvitation
+	for rows.Next() {
+		i, err := scanCollaboratorInvitation(rows)
+		if err != nil {
+			return nil, err
+		}
+		is = append(is, i)
+	}
+
+	return is, nil
+}
+
+func (s *CollaboratorService) SendInvitation(
+	appID string,
+	inviteeEmail string,
+) (*model.CollaboratorInvitation, error) {
+	sessionInfo := session.GetValidSessionInfo(s.Context)
+	if sessionInfo == nil {
+		return nil, ErrCollaboratorUnauthorized
+	}
+	invitedBy := sessionInfo.UserID
+
+	// FIXME(collaborator): Check if the invitee is collaborator
+	// It is impossible now because Admin API does not have getUserByClaim
+
+	// FIXME(collaborator): Check if the actor is a collaborator
+	// We should do this for all querys and mutations.
+
+	// Check if the invitee has a pending invitation already.
+	invitations, err := s.ListInvitations(appID)
+	if err != nil {
+		return nil, err
+	}
+	for _, i := range invitations {
+		if i.InviteeEmail == inviteeEmail {
+			return nil, ErrCollaboratorInvitationDuplicate
+		}
+	}
+
+	code := generateCollaboratorInvitationCode()
+	now := s.Clock.NowUTC()
+	// Expire in 3 days.
+	expireAt := now.Add(3 * 24 * time.Hour)
+
+	i := &model.CollaboratorInvitation{
+		ID:           uuid.New(),
+		AppID:        appID,
+		InvitedBy:    invitedBy,
+		InviteeEmail: inviteeEmail,
+		Code:         code,
+		CreatedAt:    now,
+		ExpireAt:     expireAt,
+	}
+
+	err = s.createCollaboratorInvitation(i)
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME(collaborator): Send the invitation email
+
+	return i, nil
+}
+
+func (s *CollaboratorService) GetInvitation(id string) (*model.CollaboratorInvitation, error) {
+	q := s.selectCollaboratorInvitation().Where("id = ?", id)
+	row, err := s.SQLExecutor.QueryRowWith(q)
+	if err != nil {
+		return nil, err
+	}
+	return scanCollaboratorInvitation(row)
+}
+
+func (s *CollaboratorService) DeleteInvitation(i *model.CollaboratorInvitation) error {
+	_, err := s.SQLExecutor.ExecWith(s.SQLBuilder.
+		Delete(s.SQLBuilder.FullTableName("app_collaborator_invitation")).
+		Where("id = ?", i.ID),
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *CollaboratorService) AcceptInvitation(code string) (*model.Collaborator, error) {
+	sessionInfo := session.GetValidSessionInfo(s.Context)
+	if sessionInfo == nil {
+		return nil, ErrCollaboratorUnauthorized
+	}
+	actorID := sessionInfo.UserID
+
+	now := s.Clock.NowUTC()
+	q := s.selectCollaboratorInvitation().Where("code = ? AND expire_at > ?", code, now)
+	rows, err := s.SQLExecutor.QueryWith(q)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var is []*model.CollaboratorInvitation
+	for rows.Next() {
+		i, err := scanCollaboratorInvitation(rows)
+		if err != nil {
+			return nil, err
+		}
+		is = append(is, i)
+	}
+
+	if len(is) <= 0 {
+		return nil, ErrCollaboratorInvitationInvalidCode
+	}
+
+	if len(is) > 1 {
```

Comment:
This seems strange, since we have no unique constraint on code, so an unacceptable invitation may be created. We may want to add code to the uniqueness constraint as well?

Reply:
OK.